src/main/java/org/eclipse/tractusx/managedidentitywallets/controller/SecureTokenController.java

```
@@ -0,0 +1,125 @@
+    @SecureTokenControllerApiDoc.PostSecureTokenDoc
+    public ResponseEntity<StsTokenResponse> token(
+            @Valid @RequestBody SecureTokenRequest secureTokenRequest
+    ) {
+        // handle idp authorization
+        IdpTokenResponse idpResponse = idpAuthorization.fromSecureTokenRequest(secureTokenRequest);
+        BusinessPartnerNumber bpn = idpResponse.bpn();
+        DID selfDid = new DID(walletRepo.getByBpn(bpn.toString()).getDid());
+        DID partnerDid;
+        if (Pattern.compile(StringPool.BPN_NUMBER_REGEX).matcher(secureTokenRequest.getAudience()).matches()) {
+            partnerDid = new DID(walletRepo.getByBpn(secureTokenRequest.getAudience()).getDid());
+        } else if (StringUtils.startsWith(secureTokenRequest.getAudience(), "did:")) {
+            partnerDid = new DID(secureTokenRequest.getAudience());
+        } else {
+            throw new InvalidSecureTokenRequest("You must provide an audience either as a BPN or DID.");
+        }
+
+        // create the SI token and put/create the access_token inside
+        JWT responseJwt;
+        if (secureTokenRequest.assertValidWithAccessToken()) {
+            log.debug("Signing si token.");
+            responseJwt = tokenService.issueToken(
+                    selfDid,
+                    partnerDid,
+                    JWTParser.parse(secureTokenRequest.getAccessToken())
+            );
+        } else if (secureTokenRequest.assertValidWithScopes()) {
+            log.debug("Creating access token and signing si token.");
+            responseJwt = tokenService.issueToken(
+                    selfDid,
+                    partnerDid,
+                    Set.of(secureTokenRequest.getBearerAccessScope())
+            );
+        } else {
+            throw new InvalidSecureTokenRequest("The provided data could not be used to create and sign a token.");
+        }
```

Maybe return an error if both parameters all set? I think in this special case we cannot tell what the user intended and might return the wrong token.

---

Yes, we're throwing an exception from the `assert*` methods, which will note that both parameters are set. The `InvalidSecureTokenRequest`, would be thrown if something within the provided information is invalid. Or do you mean something else?

---

This exception is probably unnecessary now, since the validator checks for consistency in the `SecureTokenRequest`, but I've decided to leave it in, just in case I've missed something.